### PR TITLE
operational certificates in shelley ledger spec

### PR DESCRIPTION
**Pull request by JaredCorduan**
This PR introduces operational certificates to the Shelley ledger spec, as described in section `4.4.7` of the delegation design document. #338 

The block header now includes an additional item, the `OCert`, defined as:

ocert

The hot keys used in the operational certs use key evolving signatures, and require an additional table in the cryptography section.  Essentially, pubkeys are fixed as before, but private keys can be evolved, with the new constraint that when you verify a signature, the verifier verifies against a minimal number of evolutions that must have occurred:

kes

The previous rule `BHEAD` has been modified so that instead of checking that the pool operator key (ie cold key) signed the header, we now check that the (hot) operational key signed the header, given the proper number of evolutions.  The delegation design document specifies that these keys must be evolved once a day, so we have introduced a new constant for the number of slots in a day, namely `SlotsPerDay`.  We can therefore enforce that the key is properly evolved by counting the days since the pool registration cert was accepted.

`BHead` now does these checks:

bhead

There is now a new transition system `OCert` which checks the validity of an operational certificate.

We need to keep track of the the counters for each stake pool's operational certificates. This is now done in `PState`, in the new variable `cs`.  Here is the new transition `OCert`:

ocertTrs

It checks that the cert is within 90 days of its start day (as specified in the deleg design doc), that the counter does not decrease, and that the cold key signed the relevant information.

### CODE REVIEW

File: latex/chain.tex [hunk 1]
```diff
@@ -409,6 +427,59 @@ \subsection{Update Nonces Transition}
...
+      \end{array}
+      \right)
+      \trans{ocert}{\var{bhb}}
+      \left(
+      \begin{array}{rcl}
+        \var{stpools} \\
+        \var{poolParams} \\
+        \var{retiring} \\
+        \var{avgs} \\
+        \varUpdate{\var{cs}\unionoverrideRight\{\var{vk}\mapsto n\}} \\
```
Comment (mgudemann) [inline latex/chain.tex]: I think it should be 
```suggestion
        \varUpdate{\var{cs}\unionoverrideRight\{\var{vk_{cold}}\mapsto n\}} \\
```

File: latex/delegation.tex [hunk 2]
```diff
@@ -507,7 +512,9 @@ \subsection{Stake Pool Rules}
         \varUpdate{\var{poolParams}} & \varUpdate{\union}
                                     & \varUpdate{\{\var{hk} \mapsto \poolParam{c}\}} \\
        \var{retiring} \\
-       \var{avgs}
+       \var{avgs} \\
+       \varUpdate{\var{cs}} & \varUpdate{\union}
+                            & \varUpdate{\{\var{hk} \mapsto 0\}} \\
```
Comment (mgudemann) [inline latex/delegation.tex]: in `OCERT` we use `VKey` and not its hash, should be done here, too

**Comment by JaredCorduan** [comment]
> I guess that a "day" will always have the same number of slots? Taking into account that real days have different length (leap second, DST changes etc.) should we play it safe and state explicitly that it is a logical day? Or maybe even use another word, how about "term" ?

That's an excellent point.  I'm not sure what the best name is, you are right that `Day` could be misleading. How about `LDay` for logical day?

File: latex/delegation.tex (hunk 2 above)
Comment (JaredCorduan) [reply latex/delegation.tex]: good catch! I think for consistency's sake we can use HashKey in OCert, so that the PStake keys are all hashkeys.

Comment (JaredCorduan) [reply latex/delegation.tex]: just pushed a change to use hashkey in ocert

**Pull request merged**